bot/core/bot.py

```diff
@@ -12,31 +17,47 @@ def __init__(self, extensions: list, *args, **kwargs) -> None:
         self.session = aiohttp.ClientSession()
 
         self.extension_list = extensions
-        self.first_on_ready = True
-
-    def run(self, token: str) -> None:
-        if not token:
-            logger.error("Missing Bot Token!")
-        else:
-            super().run(token)
+        self.initial_call = True
+
+    async def load_extensions(self) -> None:
+        """Load all listed cogs."""
+        for extension in self.extension_list:
+            try:
+                self.load_extension(extension)
+                logger.debug(f"Cog {extension} loaded.")
+            except Exception as e:
+                logger.error(f"Cog {extension} failed to load with {type(e)}: {e}")
+
+    async def db_connect(self) -> None:
+        """Estabolish connection with the database."""
+        self.database = Database(**config.DATABASE)
+        connected = await self.database.connect()
+        while not connected:
+            logger.warning("Retrying to connect to database in 5s")
+            # Synchronous sleep function to stop everything until db is connecting
+            time.sleep(5)
+            connected = await self.database.connect()
 
     async def on_ready(self) -> None:
-        """Initialize some stuff once the bot is ready."""
-        if self.first_on_ready:
-            self.first_on_ready = False
+        if self.initial_call:
+            self.initial_call = False
 
-            # Load all extensions
-            for extension in self.extension_list:
-                with logger.catch(message=f"Cog {extension} failed to load"):
-                    self.load_extension(extension)
-                    logger.debug(f"Cog {extension} loaded.")
+            await self.db_connect()
+            await self.load_extensions()
 
             logger.info("Bot is ready")
         else:
             logger.info("Bot connection reinitialized")
 
+    def run(self, token: str) -> None:
+        if not token:
+            logger.error("Missing Bot Token!")
```

Comment:
I would suggest making an exception like `MissingTokenError` and raise it, then handle in the `on_command_error` handler event

Reply:
In this case, doing this seems like a better option since it will only log the error and won't hang the `run` function which means the program will end without anything (although it won't go to close function which is admittedly a problem).

We could do something like this but I still wouldn't make a whole new exception which will only ever be used once in the whole codebase (here). Maybe a simple general `RuntimeError` would do but as I said, I quite like the way this works except the fact that it doesn't go to close function which would need an async call. This means both `aiohttp` and database won't be closed when this happens in which case it might really be better to use something like this.

I'm not quite sure if it would even be possible through `on_command_error` implies (from the name) that it handles errors which happen on commands, not in bot initialization itself so I don't think this would be caught anyway and so I'm not sure how to handle this in a way which would properly satisfy the async close.

Note that I might be wrong about the `on_command_error` and it might actually catch it but I don't think it will.

Any suggestions are welcome.